src/app/GroupLayer.tsx

```diff
@@ -0,0 +1,272 @@
+          dispatch({
+            type: "CREATE_ANNOTATION_FROM_TOKENS",
+            payload: {
+              ui: {
+                id,
+                backgroundColor: "rgb(36, 148, 178, 0.4)",
+                border: "3px solid rgb(36, 148, 178)",
+                type: "GROUP_LABEL",
+              },
+              tokens: creationState.tokens,
+            },
+          });
+          dispatch({
+            type: "CREATE_LABEL_RELATION",
+            payload: { from: Object.keys(selectedAnnotations)[0], to: id },
+          });
+          resetCreationState();
+          // As soon as a label is created, we switch user to the select tool.
+          dispatch({
+            type: "CHANGE_TOOL",
+            payload: "SELECT",
+          });
+          // After label is created we no longer want annotation to be selected.
+          // FIXME: Ask Josh, are actions dispatched in the same order ? Because if we currently
+          // selected annotation gets deselected before then we would not be able to
+          // create relation.
+          dispatch({ type: "DESELECT_ALL_ANNOTATION" });
+        },
+      };
+    }
+    case "SELECT": {
+      return {
+        cursor: "auto",
+        creationState,
+        container,
+        onMouseMove: NO_OP,
+        onMouseUp: NO_OP,
+        onMouseDown: NO_OP,
+        onMouseLeave: NO_OP,
+        onClick: () => {
+          if (tool === "SELECT") {
+            dispatch({ type: "DESELECT_ALL_ANNOTATION" });
+          }
+        },
+      };
+    }
+  }
+};
+
+export const GroupLayerAnnotation: React.FC<{
+  annotationProps: AnnotationProps;
+  annotationRef: React.MutableRefObject<HTMLDivElement | null> | undefined;
+}> = ({ annotationProps, annotationRef }) => {
+  const [tool, selectedAnnotations, annotations] = useSelector((state) => [
+    state.tool,
+    state.selectedAnnotations,
+    state.annotations,
+  ]);
+  const dispatch = useDispatch();
+  const { id, type, ...cssProps } = annotationProps;
+  const css = {
+    ...cssProps,
+    position: "absolute" as const,
+  };
+  switch (tool) {
+    case "CREATE": {
+      return (
+        <TranslucentBox
+          id={id}
+          nodeRef={annotationRef}
+          css={{ cursor: "inherit", ...css }}></TranslucentBox>
+      );
+    }
+    case "SELECT": {
+      const isSelected = Boolean(selectedAnnotations[annotationProps.id]);
+      const isFirstSelection =
+        Object.keys(selectedAnnotations)[0] === annotationProps.id;
+      return (
+        <TranslucentBox
+          id={id}
+          css={{
+            cursor: "pointer",
+            ...css,
+            border: isSelected ? "2px solid black" : css.border,
+          }}
+          onClick={(e: React.MouseEvent<HTMLElement>) => {
+            e.stopPropagation();
+            const shiftNotPressed = !e.shiftKey;
+            if (shiftNotPressed) {
+              dispatch({ type: "DESELECT_ALL_ANNOTATION" });
+            }
+            if (isSelected) {
```

Comment:
This code in the `onClick` is really clean and nice! When we need to have different things happen, having different calls to `dispatch` makes a lot of sense. However, if one action needs to trigger several actions simultaneously, creating one big new action seems like a somewhat simpler solution.